fraud/bad_encoding.go

```diff
@@ -0,0 +1,179 @@
+package fraud
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/tendermint/tendermint/pkg/consts"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// Shares contains all shares from row/col
+	// Shares that did not pass verification in rmst2d will be nil
+	// For non-nil shares MerkleProofs are computed
+	Shares []*Share
+	// Position represents the number of row/col where ErrByzantineRow/ErrByzantineColl occurred
+	Position uint8
+	isRow    bool
+}
+
+func CreateBadEncodingFraudProof(
+	height uint64,
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	errShares [][]byte,
+) (Proof, error) {
+	var tree ErasuredNamespacedMerkleTree
+	shares := make([]*Share, len(errShares))
+
+	for index, share := range errShares {
+		if share != nil {
+			tree = NewErasuredNamespacedMerkleTree(uint64(len(errShares) / 2))
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+			if err := buildTreeFromDataRoot(&tree, data, index); err != nil {
+				continue
+			}
+
+			proof, err := tree.InclusionProof(position)
+			if err != nil {
+				return nil, err
+			}
+			/*
+				NamespaceID sholud be replaced with ParitySharesNamespaceID for
+				all shares except Q0(all erasure coded shares).
+				For EDS 4x4 Q0 is [[0;0],[0;1],[1:0],[1;1]].
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
+			if index >= len(errShares)/2 || int(position) >= len(errShares)/2 {
+				namespaceID = consts.ParitySharesNamespaceID
+			}
```

Comment:
Q: While the `ErasuredNamespacedMerkleTree` does handle that internally, we manually have to do this again here because we need the exact shares that were pushed as leaves into the tree, right?

Reply:
Yes